src/api/callback.cc

```diff
@@ -66,7 +66,7 @@ InternalCallbackScope::InternalCallbackScope(Environment* env,
   if (asyncContext.async_id != 0 && !skip_hooks_) {
     // No need to check a return value because the application will exit if
     // an exception occurs.
-    AsyncWrap::EmitBefore(env, asyncContext.async_id);
+    AsyncWrap::EmitBefore(env, asyncContext.async_id, object);
```

Comment:
What if `object` is empty? (Grep around for `kAllowEmptyResource` for more on that)

Reply:
The only place `kAllowEmptyResource` is used without `kSkipAsyncHooks` is [PerIsolatePlatformData::RunForegroundTask](https://github.com/nodejs/node/blob/master/src/node_platform.cc#L386-L397), and it seems to still pass all the tests if I add it. I'm not sure what the reason is for that particular callback scope to _not_ skip async hooks, but if there _is_ a reason, we're not adequately testing for it. :thinking:

Reply:
I don't want to mark this resolved without a reply, but I'm _fairly_ sure it doesn't matter if the object is empty.

Reply:
I think some native addon could pass an empty receiver to `MakeCallback`.

Reply:
@Flarna I’ve looked at it before writing that comment, but I think `MakeCallback` actually forbids this? I was surprised too.

@Qard I think the current code would handle it fine because `GetReturnValue().Set()` silently ignores empty values. That’s okay, but I’d like to eventually deal with this in a better way.

Reply:
correct, it is [forbidden](https://github.com/nodejs/node/blob/master/src/api/callback.cc#L147)

There could be a mismatch between the given receiver and the actual async resource object and there could be also reuse of the receiver but that's another topic.

Reply:
So what's the next step? Do you want me to find a fix for this before landing this? Or can we deal with that separately?

Reply:
Sorry to be not clear regarding this. I think we can move this to a seperate PR.

Reply:
Marking as resolved as it seems like there's nothing to do immediately in the PR.